### PR TITLE
fix(snapshot): fix memleak in lv_snapshot

### PR DESCRIPTION
**Pull request by terry0012**
### Description of the feature or fix

When we take a snapshot for a lv_obj, if there is a `LV_LAYER_TYPE_TRANSFORM`  object in the children.
A new layer named `new_layer` will be malloced in `refr_obj` function.
But it will not be freed after `lv_snapshot_take_to_buf` done as we call `lv_draw_dispatch_layer(NULL, &layer);` 

Because the `new_layer` is freed  only when `disp` is not NULL as the below code.
```
                /*Remove the layer from  the display's*/
                if(disp) {
                    lv_layer_t * l2 = disp->layer_head;
                    while(l2) {
                        if(l2->next == layer_drawn) {
                            l2->next = layer_drawn->next;
                            break;
                        }
                        l2 = l2->next;
                    }

                    if(disp->layer_deinit) disp->layer_deinit(disp, layer_drawn);
                    lv_free(layer_drawn);
                }
```
### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.12](https://github.com/szepeviktor/astyle/releases/tag/v3.4.12) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Comment by kisvegabor** [comment]
Wonderful, thank you! 

Could you add a [test](https://github.com/lvgl/lvgl/blob/master/tests/README.md) too? Having a [screenshot](https://github.com/lvgl/lvgl/blob/master/tests/README.md#asserts) for a transformed layer and an assert for the [memory leak](https://github.com/lvgl/lvgl/blob/master/tests/src/test_cases/cache/test_cache.c#L131-L138) would be great.

**Comment by terry0012** [comment]
> Wonderful, thank you!
> 
> Could you add a [test](https://github.com/lvgl/lvgl/blob/master/tests/README.md) too? Having a [screenshot](https://github.com/lvgl/lvgl/blob/master/tests/README.md#asserts) for a transformed layer and an assert for the [memory leak](https://github.com/lvgl/lvgl/blob/master/tests/src/test_cases/cache/test_cache.c#L131-L138) would be great.

OK. 
I add a transform obj under active screen in function `test_snapshot_should_not_leak_memory`.
And test_snapshot can detect the memory leak without this patch.

By the way, I found `test_snapshot` will stuck in `LVGL_CI_USING_SYS_HEAP` mode if `lv_obj_set_style_transform_rotation` is added.
I will continue to check it further.

**Comment by kisvegabor** [comment]
I've checked it an it seems it really stuck in `lv_snapshot_take` only when `lv_obj_set_style_transform_rotation` is enabled and only in the tests. Running it normally  (not in tests) works well even with the very same lv_conf.h as in the tests.

**Comment by terry0012** [comment]
> I've checked it an it seems it really stuck in `lv_snapshot_take` only when `lv_obj_set_style_transform_rotation` is enabled and only in the tests. Running it normally  (not in tests) works well even with the very same lv_conf.h as in the tests. 
> 

Yes. it's a dead lock problem.

**Comment by terry0012** [comment]
In the test case, we have three draw task:

- draw rect for screen obj on snapshot layer (`Task 1`)
- draw label for label obj on a transform layer (`Task 2`)
- draw transforme layer on snapshot layer (`Task 3`)

For `Task 3`, the init state is `LV_DRAW_TASK_STATE_WAITING` as it need to wait `Task 2` done.
But when we call `lv_draw_dispatch_layer` , it only  traverse the snapshot layer to change the task state.

At last, when `lv_draw_dispatch_wait_for_request` called, it'll call `lv_draw_get_next_available_task`  to search a draw task, but `Task 3` is not available because it is in `LV_DRAW_TASK_STATE_WAITING` state.

At this moment,  `layer.draw_task_head` is not NULL , `LVGL` thread is waiting a signal, `Render` thread is also waiting a signal as `u->task_act` is NULL.  Then, a dead lock happened.

File: tests/src/test_cases/test_snapshot.c [hunk 1]
```diff
@@ -30,6 +35,7 @@ void test_snapshot_should_not_leak_memory(void)
 
     lv_mem_monitor(&monitor);
     final_available_memory = monitor.free_size;
+    lv_obj_delete(label);
 
     TEST_ASSERT_EQUAL(initial_available_memory, final_available_memory);
```
Comment (kisvegabor) [inline tests/src/test_cases/test_snapshot.c:67]: Please add a separate screen shot test too for transformations, just to be sure.

File: tests/src/test_cases/test_snapshot.c [hunk 2]
```diff
@@ -30,6 +35,7 @@ void test_snapshot_should_not_leak_memory(void)
```
Comment (kisvegabor) [inline tests/src/test_cases/test_snapshot.c:62]: `NUM_SNAPSHOTS` is `1` now. Please set it to something larger, e.g. 10.

**Review by kisvegabor — COMMENTED** [review]
Thank you for the investigation! I see no issue with running a full dispatching here.

File: tests/src/test_cases/test_snapshot.c (hunk 1 above)
Comment (kisvegabor) [reply tests/src/test_cases/test_snapshot.c:67]: Where the screen shot test is added?

Comment (terry0012) [reply tests/src/test_cases/test_snapshot.c:67]: `test_snapshot_with_transform_should_not_leak_memory` function is the snapshot test for transformation.

Comment (kisvegabor) [reply tests/src/test_cases/test_snapshot.c:67]: But there is no  ` TEST_ASSERT_EQUAL_SCREENSHOT` in it so we can't see if a transformed widget was rendered correctly.

Comment (terry0012) [reply tests/src/test_cases/test_snapshot.c:67]: > But there is no ` TEST_ASSERT_EQUAL_SCREENSHOT` in it so we can't see if a transformed widget was rendered correctly.

got it.

**Review by kisvegabor — APPROVED** [review]
Thank you!

**Review by FASTSHIFT — APPROVED** [review]

**Pull request merged**